src/rust/rust-proto-new/src/graplinc/grapl/api/pipeline_ingress/v1beta1.rs

```diff
@@ -137,3 +137,199 @@ impl SerDe for PublishRawLogsResponse {
         response_proto.try_into()
     }
 }
+
+//
+// client
+//
+
+pub mod client {
+    use crate::{
+        graplinc::grapl::api::pipeline_ingress::v1beta1::{
+            PublishRawLogsRequest,
+            PublishRawLogsResponse,
+        },
+        protobufs::graplinc::grapl::api::pipeline_ingress::v1beta1::pipeline_ingress_service_client::PipelineIngressServiceClient as PipelineIngressServiceClientProto,
+    };
+
+    use thiserror::Error;
+
+    #[derive(Debug, Error)]
+    pub enum PublishRawLogsError {
+        #[error("todo")]
+        Todo,
+    }
+
+    pub struct PipelineIngressClient {
+        proto_client: PipelineIngressServiceClientProto<tonic::transport::Channel>
+    }
+
+    impl PipelineIngressClient {
+        pub async fn publish_raw_logs(
+            &self,
+            raw_log: PublishRawLogsRequest
+        ) -> Result<PublishRawLogsResponse, PublishRawLogsError> {
+            todo!()
+        }
+    }
+}
+
+//
+// server
+//
+
+/// This module contains the gRPC server for serving the pipeline ingress
+/// API. We encapsulate all the types generated by the protocol buffer compiler
+/// and instead expose our own "sanitized" version of the API. Users should
+/// implement their business logic by constructing an object conforming to the
+/// PipelineIngressApi trait. To run the gRPC server implementing that business
+/// logic, inject the PipelineIngressApi implementation into the
+/// PipelineIngressServer's constructor.
+pub mod server {
+    use std::{net::SocketAddr, marker::PhantomData};
+
+    use crate::{
+        graplinc::grapl::api::pipeline_ingress::v1beta1::{
+            PublishRawLogsRequest,
+            PublishRawLogsResponse,
+        },
+        protobufs::graplinc::grapl::api::pipeline_ingress::v1beta1::{
+            pipeline_ingress_service_server::{
+                PipelineIngressService as PipelineIngressServiceProto,
+                PipelineIngressServiceServer as PipelineIngressServiceServerProto
+            },
+            PublishRawLogsRequest as PublishRawLogsRequestProto,
+            PublishRawLogsResponse as PublishRawLogsResponseProto,
+        },
+    };
+    use futures::TryFutureExt;
+    use thiserror::Error;
+    use tonic::{
+        Request,
+        Response,
+        Status,
+        transport::Server,
+    };
+
+    //
+    // protocol buffer stuff
+    //
+
+    /// This struct implements the internal gRPC representation of the pipeline
+    /// ingress server. We've implemented the service trait generated by tonic
+    /// in such a way that it delegates to an externally supplied
+    /// PipelineIngressApi. This way all the protocol buffer compiler generated
+    /// types are encapsulated, and the public API is implemented in terms of
+    /// this crate's sanitized types.
+    struct PipelineIngressProto<T, E>
+    where
+        T: PipelineIngressApi<E>,
+        E: ToString + 'static
+    {
+        api_server: T,
+        _e: PhantomData<E>
```

Comment:
`E` is used in `T`. Unfortunately, if I don't include it here I get a compiler error complaining that `E` doesn't exist. I do _need_ `E` here because it's the error type passed in via the `PipelineIngressApi<E>` interface.

Reply:
Oops, I missed that it was in T. Yep, that's correct.